Various fixes for Unix Pen and CustomLineCap.

mellinoe opened the pull request:
Pen changes
* Fix behavior of Pen.CustomEndCap/StartCap on Unix.
  * Use a cached value for CustomEndCap, because a get accessor is not available via libgdiplus.
  * Try to match the exception behavior of Windows for these two properties.
  * Correctly clone CustomLineCap objects before storing them in set_CustomEndCap / set_CustomStartCap.
  * MultiplyTransform now explicitly checks for a disposed matrix before using it. libgdiplus does not ignored null matrix handles -- it throws an exception.
  * Re-enable all Pen tests on all platforms.

CustomLineCap changes
* Stop using unavailable GdipGetCustomLineCapType function on Unix
* Allow AdjustableArrowCap objects to be successfully cloned despite the above

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Drawing2D/AdjustableArrowCap.Unix.cs:
```diff
@@ -0,0 +1,22 @@
+            int status = SafeNativeMethods.Gdip.GdipCloneCustomLineCap(new HandleRef(this, nativeCap), out clonedCap);
+
+            if (status != SafeNativeMethods.Gdip.Ok)
+                throw SafeNativeMethods.Gdip.StatusException(status);
```
@hughbe added a helper for this:
`SafeNativeMethods.Gdip.CheckStatus(status);`

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Pen.Unix.cs:
```diff
@@ -0,0 +1,80 @@
+                IntPtr lineCap = IntPtr.Zero;
+                int status = SafeNativeMethods.Gdip.GdipGetPenCustomStartCap(new HandleRef(this, NativePen), out lineCap);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+                if (lineCap == IntPtr.Zero)
```
It might return a success error code even if the arg was invalid?

mellinoe replied:
Right: libgdiplus just returns a null handle if the Pen doesn't have an associated CustomStartCap. On Windows, the accessor throws an ArgumentException if you haven't set the CustomStartCap yet.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Pen.Unix.cs:
```diff
@@ -0,0 +1,80 @@
+                // If the CustomEndCap has never been set, this accessor should throw.
+                if (_cachedEndCap == null)
+                {
+                    throw new ArgumentException(SR.Format(SR.GdiplusInvalidParameter));
```
`ArgumentException` is the right exception type for this, even though there's no argument?

mellinoe replied:
Yeah -- the exception types thrown from this library often make no sense.

mellinoe replied:
In this case, it might help to think of the native function itself, which does have a parameter (the native handle of the pen). Still, it's not much of an "argument exception" (IMO), because the parameter passed in is actually a valid Pen handle -- it just doesn't have an end cap associated with it.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Pen.Unix.cs:
```diff
@@ -0,0 +1,80 @@
+            {
+                if (_immutable)
+                {
+                    throw new ArgumentException(SR.Format(SR.CantChangeImmutableObjects, nameof(Pen)));
```
Ditto

mellinoe commented:
@akoeplinger I just wanted to point out that I think mono is currently affected by the CustomLineCap issue I fixed in here. As part of some of the code consolidation work, mono started using our version of CustomLineCap.cs, but that tries to call `GdipGetCustomLineCapType`, which isn't implemented by libgdiplus. As far as I can tell, that code path is only hit when calling `CustomLineCap.Clone()`.

The other stuff I fixed probably isn't that important for mono; it's just edge-case exception behavior type stuff.

It might be a good idea to sync up the codebases once this goes in. It will also help me make sure that I haven't made any incompatible changes for mono.

mellinoe commented:
@dotnet-bot Test OSX x64 Debug Build (Looks like the build timed out for some reason).
